bigquery_schema_generator/generate_schema.py

```diff
@@ -105,17 +105,6 @@ def __init__(
         # rest of the file.
         self.ignore_invalid_lines = ignore_invalid_lines
 
-        # 'infer_mode' is supported for only input_format = 'csv' because
-        # the header line gives us the complete list of fields to be expected in
-        # the CSV file. In JSON data files, certain fields will often be
-        # completely missing instead of being set to 'null' or "". If the field
-        # is not even present, then it becomes incredibly difficult (not
-        # impossible, but more effort than I want to expend right now) to figure
-        # out which fields are missing so that we can mark the appropriate
-        # schema entries with 'filled=False'.
-        if infer_mode and input_format != 'csv':
```

Comment:
I think this might be a bad idea. I know that you tried to recover some of this functionality in `flatten_schema_map()`, I am not sure that's sufficient. The reason is that if --infer_mode is enabled in JSON mode, it will do the wrong thing when JSON lines are missing those fields completely. All of the records will have a filled=True state, so I think the program will incorrectly infer the mode to be `REQUIRED`. This will break `bq load` when loading that dataset with this generated schema.

I realize that I was the one who suggested overloading the `--infer_mode` for the `REQUIRED->NULLABLE` transition, but I'm starting to think that we need a separate flag for this. To avoid enabling `--infer_mode` for JSON format.

Reply:
Will continue on the next PR I submit with squashed commits but why would this be bad given that added one more check to only apply this in `flatten_schema_map()` if the input_format is a CSV. If it's a JSON input format it will not elevate these to REQUIRED.